### PR TITLE
feat(janus-idp/cli): add `package package-dynamic-plugins` command

### PR DESCRIPTION
**Pull request by kadel**
This PR adds a command that packages dynamic plugins into an image container.

To package dynamic plugin execute `package-dynamic-plugins` command and specify full image name using `-t` flag.
`podman` is used by default. Alternatively you can also use `docker` by providing `--container-tool docker`

example:
```
▶ npx @janus-idp/cli package package-dynamic-plugins -t quay.io/tkral/dynamic-plugins-todo:latest
Using existing 'dist-dynamic' directory at plugins/todo
Using existing 'dist-dynamic' directory at plugins/todo-backend
Copying 'plugins/todo/dist-dynamic' to '/var/folders/5c/67drc33d0018j6qgtzqpcsbw0000gn/T/package-dynamic-pluginsQ2w8gF/backstage-community-plugin-todo
Copying 'plugins/todo-backend/dist-dynamic' to '/var/folders/5c/67drc33d0018j6qgtzqpcsbw0000gn/T/package-dynamic-pluginsQ2w8gF/backstage-community-plugin-todo-backend-dynamic
Writing plugin registry metadata to '/var/folders/5c/67drc33d0018j6qgtzqpcsbw0000gn/T/package-dynamic-pluginsQ2w8gF/index.json'
Creating image using podman
  executing     echo "from scratch
COPY . .
" | podman build --annotation com.redhat.rhdh.plugins='[{"backstage-community-plugin-todo":{"name":"@backstage-community/plugin-todo","version":"0.2.40","description":"A Backstage plugin that lets you browse TODO comments in your source code","backstage":{"role":"frontend-plugin","pluginId":"todo","pluginPackages":["@backstage-community/plugin-todo","@backstage-community/plugin-todo-backend"]},"homepage":"https://backstage.io","repository":{"type":"git","url":"https://github.com/backstage/community-plugins","directory":"workspaces/todo/plugins/todo"},"license":"Apache-2.0"}},{"backstage-community-plugin-todo-backend-dynamic":{"name":"@backstage-community/plugin-todo-backend","version":"0.3.19","description":"A Backstage backend plugin that lets you browse TODO comments in your source code","backstage":{"role":"backend-plugin","pluginId":"todo","pluginPackages":["@backstage-community/plugin-todo","@backstage-community/plugin-todo-backend"]},"homepage":"https://backstage.io","repository":{"type":"git","url":"https://github.com/backstage/community-plugins","directory":"workspaces/todo/plugins/todo-backend"},"license":"Apache-2.0"}}]' -t 'quay.io/tkral/dynamic-plugins-todo:latest' -f - .
     ✔
Successfully built image quay.io/tkral/dynamic-plugins-todo:latest with following plugins:
  backstage-community-plugin-todo
  backstage-community-plugin-todo-backend-dynamic

Configuration example for the dynamic-plugins.yaml:

packages:
- package: oci://quay.io/tkral/dynamic-plugins-todo:latest!backstage-community-plugin-todo
  disabled: false
  pluginConfig:
    dynamicPlugins:
      # add required plugin configuration here

- package: oci://quay.io/tkral/dynamic-plugins-todo:latest!backstage-community-plugin-todo-backend-dynamic
  disabled: false
  pluginConfig:
    dynamicPlugins:
      # add required plugin configuration here
```

After pushing image to the registry you can load it by https://github.com/janus-idp/backstage-showcase/pull/1479

fixes https://issues.redhat.com/browse/RHIDP-3571

### CODE REVIEW

File: packages/cli/src/commands/index.ts [hunk 1]
```diff
@@ -139,6 +139,31 @@ export function registerScriptCommand(program: Command) {
...
+      '--preserve-temp-dir',
+      'Leave the temporary staging directory on the filesystem instead of deleting it',
+    )
+    .requiredOption(
+      '-t, --tag <tag>',
+      'Tag name to use when building the plugin registry image',
+    )
+    .option(
+      '--container-tool <tool>',
+      'The container tool to use for building the image (podman or docker, podman is the default)',
```
Comment (gashcrumb) [inline packages/cli/src/commands/index.ts:161]: Am I missing where the default is specified in this PR?  Isn't there another argument to `option` needed here to specify that default value?  Apologies if I missed this elsewhere.

Comment (kadel) [reply packages/cli/src/commands/index.ts:161]: oh, you are not missing it. I'm the one who missed it :-) 

I forgot to push updated commit 😇

File: packages/cli/src/commands/package-dynamic-plugins/command.ts [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    const { packageDirectory, packageFilePath } = pluginPkg;
+    if (
+      !fs.existsSync(path.join(packageDirectory, 'dist-dynamic')) ||
+      forceExport
+    ) {
+      Task.log(
+        `Running export-dynamic-plugin on plugin package ${packageFilePath}`,
+      );
+      try {
+        await Task.forCommand(`yarn export-dynamic`, { cwd: packageDirectory });
```
Comment (jesuino) [inline packages/cli/src/commands/package-dynamic-plugins/command.ts:48]: Shouldn't we have different options for frontend/backend plugins?( `--no-in-place` for frontend or `--embed-as-dependencies` for backend)

File: packages/cli/src/commands/package-dynamic-plugins/command.ts [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    await fs.writeFile(
+      metadataFile,
+      JSON.stringify(pluginRegistryMetadata, undefined, 2),
+    );
+    // run the command to generate the image
+    Task.log(`Creating image using ${containerTool}`);
+    await Task.forCommand(
+      `echo "from scratch
+COPY . .
+" | ${containerTool} build --annotation com.redhat.rhdh.plugins='${JSON.stringify(pluginRegistryMetadata)}' -t '${tag}' -f - .
```
Comment (jesuino) [inline packages/cli/src/commands/package-dynamic-plugins/command.ts:131]: Is there a way to check if the containerTool is available before building plugins? A fail fast approach: "The tool containerTool is not installed, please make sure it is installed before running this command".

File: packages/cli/src/commands/index.ts [hunk 4]
```diff
@@ -139,6 +139,32 @@ export function registerScriptCommand(program: Command) {
...
```
Comment (jesuino) [inline packages/cli/src/commands/index.ts:161]: Would it be possible to limit it for the supported tools only and avoid user to try something that we do not expect.

**Review by jesuino — COMMENTED** [review]
Thanks for this great command. Here are some suggested changed/questions

File: packages/cli/src/commands/package-dynamic-plugins/command.ts (hunk 2 above)
Comment (jesuino) [reply packages/cli/src/commands/package-dynamic-plugins/command.ts:48]: I see now, we call the user defined command, hence user must be responsible to add additional flags

**Review by jesuino — APPROVED** [review]
Thanks for the changes. LGTM!

**Review by gashcrumb — APPROVED** [review]
:+1:

**Pull request merged**